File: kalshi_api.R
library(httr2)
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(tibble)

source("helpers.R")

get_portfolio_balance <- function() {
  base_url <- "https://api.elections.kalshi.com"
   
  path <- "/trade-api/v2/portfolio/balance"
  headers <- create_headers("GET", path)
  req <- http_build(base_url, path, headers)
  resp <- http_run(req)
  
  balance <- resp$balance / 100
  return(balance)
}

get_events <- function(limit = 100, with_nested_markets = TRUE) {
  base_url <- "https://api.elections.kalshi.com"
  
  path <- "/trade-api/v2/events"
  headers <- create_headers("GET", path)
  query_params <- list(limit = limit, with_nested_markets = with_nested_markets)
  req <- http_build(base_url, path, headers, query_params)
  resp <- http_run(req)
  
  events <- map(resp$events, tibble::as_tibble)
  events <- 
    events |> bind_rows() |> 
    select(-c(event_ticker, title, category)) |> 
    unnest_wider(markets)
  return(events)
}

get_event <- function(event_ticker, with_nested_markets = TRUE) {
  base_url <- "https://api.elections.kalshi.com"
  
  path <- str_glue("/trade-api/v2/events/{event_ticker}")
  headers <- create_headers("GET", path)
  query_params <- list(with_nested_markets = with_nested_markets)
  req <- http_build(base_url, path, headers, query_params)
  resp <- http_run(req)
  
  event <- 
    as_tibble(resp$event) |> 
    # TODO: probably am dropping the wrong title
    select(-c(event_ticker, title, category)) |> 
    unnest_wider(markets)
  return(event)
}

resp <- get_event("KXCABCOUNT-25MAR01", with_nested_markets = TRUE)

next_req <- function(resp, req) {
    cursor <- resp_body_json(resp)$cursor
    if (cursor == "")
      return(NULL)
    req |> req_url_query(cursor = cursor)
}

get_trades <- function(limit = 100,
                       market_ticker) {
  base_url <- "https://api.elections.kalshi.com"
  
  path <- str_glue("/trade-api/v2/markets/trades")
  headers <- create_headers("GET", path)
  query_params <- list(
    limit = limit,
    ticker = market_ticker
  )
  
  req <- http_build(base_url, path, headers, query_params)
  resp <- req_perform_iterative(
    req,
    next_req,
    max_reqs = Inf
  )
  trades <- resps_data(resp, function(resp) {
    data <- resp_body_json(resp)$trades
    map(data, as_tibble)
  }) |> 
    bind_rows()
  return(trades)
} 

resp <- get_trades(1000, "KXCABCOUNT-25MAR01-15")

resp |> 
  mutate(created_time = lubridate::as_datetime(created_time)) |>
  
  

File: kalshi_api.jl
using HTTP
using Base64
using StringEncodings
using JSON3
using DataFrames

include("helpers.jl")

function get_portfolio_balance()
    base_url = "https://api.elections.kalshi.com"

    path = "/trade-api/v2/portfolio/balance"
    headers = create_headers("GET", path)
    resp = HTTP.get(base_url * path; headers=headers)
    portfolio = JSON3.read(decode(resp.body, "UTF-8"))
    return portfolio[:balance] / 100
end

function get_events(limit::Int = 100, with_nested_markets::Bool = true, all_events::Bool = true)
    base_url = "https://api.elections.kalshi.com"
    path = "/trade-api/v2/events"
    headers = create_headers("GET", path)
    query_params = Dict{String, Any}("limit" => limit, "with_nested_markets" => with_nested_markets) 

    resp = HTTP.get(base_url * path; headers = headers, query = query_params)
    events = JSON3.read(decode(resp.body, "UTF-8"))

    if all_events
        data = paginate(base_url, path, headers, query_params, events, :events)
        return data
    end
    return events
end


function get_event(event_ticker::String, with_nested_markets::Bool = true)
    base_url = "https://api.elections.kalshi.com"

    path = "/trade-api/v2/events/$event_ticker"
    headers = create_headers("GET", path)
    query_params = Dict("with_nested_markets" => with_nested_markets)
    resp = HTTP.get(base_url * path; headers = headers, query = query_params)
    event = JSON3.read(decode(resp.body, "UTF-8"))
    return event[:event]
end


function get_trades(market_ticker::String, limit::Int = 1000, all_trades::Bool = true)::DataFrame
    base_url = "https://api.elections.kalshi.com"
    path = "/trade-api/v2/markets/trades"
    headers = create_headers("GET", path)
    query_params = Dict{String, Any}("limit" => limit, "ticker" => market_ticker)

    resp = HTTP.get(base_url * path; headers = headers, query = query_params)
    trades = JSON3.read(decode(resp.body, "UTF-8"))

    if all_trades
        data = paginate(base_url, path, headers, query_params, trades, :trades)
        data = DataFrame(reduce(vcat, data))
        return data
    end
    return DataFrame(trades)
end

function get_orderbook(market_ticker::String, depth::Int = 99)
    base_url = "https://api.elections.kalshi.com"
    path = "/trade-api/v2/markets/$market_ticker/orderbook"
    headers = create_headers("GET", path)
    query_params = Dict("depth" => depth)

    resp = HTTP.get(base_url * path; headers = headers, query = query_params)
    orderbook = JSON3.read(decode(resp.body, "UTF-8"))
    return orderbook[:orderbook]
end

function sort_orderbook(orderbook)
    orderbook = sort(collect(orderbook), by = x -> x[1])
    return orderbook
end

function separate_orderbook(orderbook)
    yes_side = orderbook[:yes]
    no_side = orderbook[:no]

    yes_side_price = map(x -> x[1], yes_side)
    yes_side_contracts = map(x -> x[2], yes_side)

    no_side_price = map(x -> x[1], no_side)
    no_side_contracts = map(x -> x[2], no_side)

    yes_side = sort_orderbook(Dict(zip(yes_side_price, yes_side_contracts)))
    no_side = sort_orderbook(Dict(zip(no_side_price, no_side_contracts)))

    return yes_side, no_side
end

gallego_yes, gallego_no = separate_orderbook(gallego_orderbook)

gallego_yes
gallego_no